Issue-104 - Display local servers list before joining game

phalasz opened the pull request:
Enhancement for #104 

- Add a list of selectable local servers to the default multiplayer menu.
- Server Query protocol implementation to get server information
   - inspired by https://developer.valvesoftware.com/wiki/Server_queries and the SQP implementation in the Unity FPS Sample.
- Move all configuration to a ScriptableObject asset instead of having it spread out in several places

Roobubba commented:
On 2 64-bit windows 10 machines on a LAN, both as built executables or 1 exe + 1 unity editor (host or client), I found the following behaviour:
1) Clicking on the highlighted server didn't populate the IP/port boxes & double clicking had no effect
2) One client didn't update when the other machine hosted, but when that client was host, it *did* show up the other way round. No idea why that would be, as they used the same built files
3) When I backed out from the host by hitting escape, the server was no longer visible and updates did not occur for the other machine until the application was restarted.

Hope this helps!

phalasz commented:
Interesting. I can't reproduce points 1 and 2 on Unity 2018.3.13f1

Roobubba commented:
I'm on Unity 2018.3.0f2. I'll upgrade & check back. I did retest by making a new project and trying again and reproduced the same behaviour on this older version of unity.

phalasz commented:
Very odd. I'll keep looking to see if I can reproduce.

phalasz commented:
Hey, not sure what you mean in your point 3. The only issue I see there was a socket exception as both instances running switched back to default mode before connectiong and tried to listen on the same SQP port on the same machine. Obviously the server will not be visible as it was just closed by hitting Esc so there is nothing to show up.

phalasz commented:
Okay, fixed point 1.

theWill commented:
just trying to get this working and hitting an error:
 'MultiplayerMenu' is a namespace but is used like a type

I'm not a namespace expert but perhaps rename this new namespace to 'MultiplayerMenus' so it doesnt conflict with the MultiplayerMenu.cs class?

phalasz commented:
@Roobubba could you check again and see if your issues have been fixed?

Roobubba commented:
Checked today with Unity Editor version 2018.3.14f1

My old version 2 is probably a network problem at my end. I have observed this behaviour with separate code that is unrelated. Please discard that comment!

Issue 1 is fixed: I can now click the highlighted server and the boxes below are populated correctly. 
Issue 2 should be disregarded: I think my network is at fault. 
Issue 3 is fixed: I can now host, connect with a client, 'play', then escape back to menu, and re-host and re-connect successfully multiple times in succession with no issues.

No further issues found. Looking good!

theWill commented:
I tested the latest a bit. I like the change to use a scriptable object a lot but I think there may be confusion for people upgrading from previous forge versions. Here are the issues I encountered when upgrading an existing forge install so far:

upgrade guide:

any classes using MultiplayerMenu I had to add the following:
'using BeardedManStudios.MultiplayerMenu;'

one issue is all the settings (like Dont Change Scene, etc) previously on my MultiplayerMenu.cs script 
are lost after upgrading since those are now on the ForgeSettings scriptable object.
For helping people upgrade it might be better to keep the old settings for some time period and mark them as deprecated.

theWill commented:
other changes required for upgrade:

1) find old settings on multiplayerMenu and copy into the default forge settings script able object (or create a new scrip-table object with the settings.)

2) Any prefabs or scene objects with MulitplayerMenu need to have their ForgeSettings field filled with a reference to the above scriptable object.

3) When I upgrade I typically copy all the new filed EXCEPT the NetworkManager since it is filled in with a bunch of my custom game references. So, you also need to fill the NetworkManager prefab with an instance to the above scriptable object.

phalasz commented:
@theWill makes sense, let me make some adjustments then.

phalasz commented:
@theWill 

1. I've added back the fields so when upgrading the values should not be lost. I've also added a new inspector that should tell the user that the fields are obsolete and with the provided button they could save their settings.
2. In the new default MP menu script I've added in a check in `Awake` and log and Error and a Warning if the settings field is not set then try to find the default settings object and link that.
3. The NetworkManager script should already log an error in Awake if the field is empty and try to load the default settings from disk.

I'll get on the upgrade docs next.

theWill commented:
Nice changes. Thanks. I'll dive into testing the join menu screen as I implement it with my project. I'll let you know how it goes.

These changes are going to be an awesome upgrade.
Thanks!

NFMynster commented:
@theWill @Roobubba all good and stable?

theWill commented:
Yes, seems solid and a good upgrade for all. I'll be testing more on Android and iOS in the coming 2 weeks. @NFMynster

NFMynster commented:
Alright, waiting for further greenlight. Thank you

Roobubba commented:
I have only tested in Windows 64-bit editor and standalone builds, but in those tests it's working nicely, I didn't find any bugs.

phalasz commented:
@theWill any further feedback or are we good to merge this in?

theWill commented:
looks solid. please merge.

On Sun, Jun 23, 2019 at 6:04 AM Peter Halasz <notifications@github.com>
wrote:

> @theWill <https://github.com/theWill> any further feedback or are we good
> to merge this in?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BeardedManStudios/ForgeNetworkingRemastered/pull/266?email_source=notifications&email_token=AAIA5WTCY343YWHHOFA6MUDP35YHNA5CNFSM4HI5FPG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYK57SI#issuecomment-504750025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIA5WUCMB5UFDWZKVGAQBLP35YHNANCNFSM4HI5FPGQ>
> .
>


-- 
Sincerely,
Will Supinski

BrentFarris commented:
I have merged a big cleanup branch and so you will probably need to merge current develop into your fork and resolve conflicts, sorry about this. The main breaking change in develop is the folder paths have updated their names to not have spaces. Please see [this commit](https://github.com/BeardedManStudios/ForgeNetworkingRemastered/pull/292) for more info on the changes

BrentFarris commented:
Thanks for the work on this, I noticed the long(ish) conversation thread so I'm going on faith a bit on this one, will test all the merges when they get into develop

phalasz commented:
Thanks @BrentFarris I'll do one last check to make sure that I have all the things working still after the changes to develop and then merge this in.
